### PR TITLE
feat(rollapp): enforce EIP155 chain ID for rollapp creation

### LINKED ISSUE #1014: All rollapps id should follow the eip155 standards
Currently we only enforce it on evm rollapps.
for better standartization and integration we should enforce it on all rollapps.

### PR DESCRIPTION
**Pull request by zale144**
## Description

----

Closes #1014 

**All** items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow-up issues.

PR review checkboxes:

I have...

- [ ]  Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ]  Targeted PR against the correct branch
- [ ]  included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Linked to the GitHub issue with discussion and accepted design
- [ ]  Targets only one GitHub issue
- [ ]  Wrote unit and integration tests
- [ ]  Wrote relevant migration scripts if necessary
- [ ]  All CI checks have passed
- [ ]  Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code)
- [ ]  Updated the scripts for local run, e.g genesis_config_commands.sh if the PR changes parameters
- [ ]  Add an issue in the [e2e-tests repo](https://github.com/dymensionxyz/e2e-tests) if necessary

SDK Checklist
- [ ] Import/Export Genesis
- [ ] Registered Invariants
- [ ] Registered Events
- [ ] Updated openapi.yaml
- [ ] No usage of go `map`
- [ ] No usage of `time.Now()`
- [ ] Used fixed point arithmetic and not float arithmetic
- [ ] Avoid panicking in Begin/End block as much as possible
- [ ] No unexpected math Overflow
- [ ] Used `sendCoin` and not `SendCoins`
- [ ] Out-of-block compute is bounded
- [ ] No serialized ID at the end of store keys
- [ ] UInt to byte conversion should use BigEndian

Full security checklist [here](https://www.faulttolerant.xyz/2024-01-16-cosmos-security-1/)

----;

For Reviewer:

- [ ]  Confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Reviewers assigned
- [ ]  Confirmed all author checklist items have been addressed

---;

After reviewer approval:

- [ ]  In case the PR targets the main branch, PR should not be squash merge in order to keep meaningful git history.
- [ ]  In case the PR targets a release branch, PR must be rebased.

## Summary by CodeRabbit

- **New Features**
	- Enhanced rollapp ID generation using a new randomization method for improved uniqueness.
	- Introduced a new method for retrieving rollapps by name.

- **Bug Fixes**
	- Improved error handling in rollapp creation processes to provide clearer error messages.

- **Tests**
	- Expanded test cases to include dynamic rollapp ID generation and ensure functionality across various scenarios.

- **Chores**
	- Updated import statements for clarity and organization.

### CODE REVIEW

File: x/rollapp/keeper/msg_server_create_rollapp.go [hunk 1]
```diff
@@ -45,21 +45,20 @@ func (k msgServer) CreateRollapp(goCtx context.Context, msg *types.MsgCreateRoll
...
-	}
-	// check to see if the RollappId has been registered before with same key
+
+	// check to see if the RollappId has been registered before with same EIP155 ID
 	existingRollapp, isFound := k.GetRollappByEIP155(ctx, rollappId.GetEIP155ID())
 	// allow replacing EIP155 only when forking (previous rollapp is frozen)
 	if !isFound {
+		// if not forking, check to see if the Rollapp has been registered before with same name
+		if _, isFound = k.FindRollappByName(ctx, rollappId.GetName()); isFound {
+			return types.ErrRollappExists
```
Comment (danwt) [inline x/rollapp/keeper/msg_server_create_rollapp.go:60]: might wanna add a reason to this error

Comment (danwt) [reply x/rollapp/keeper/msg_server_create_rollapp.go:60]: with errorsmod.Wrap

Comment (danwt) [reply x/rollapp/keeper/msg_server_create_rollapp.go:60]: also the description of the ErrRollappExists specificiall refers to ID, not name, so it's misleading

Comment (danwt) [reply x/rollapp/keeper/msg_server_create_rollapp.go:60]: imo `return errorsmod.Wrap(gerrc.ErrAlreadyExists, "name")`

File: x/rollapp/keeper/msg_server_create_rollapp.go [hunk 2]
```diff
@@ -45,21 +45,20 @@ func (k msgServer) CreateRollapp(goCtx context.Context, msg *types.MsgCreateRoll
...
 	// check to see if the RollappId has been registered before
 	if _, isFound := k.GetRollapp(ctx, rollappId.GetChainID()); isFound {
 		return types.ErrRollappExists
 	}
-	if !rollappId.IsEIP155() {
-		return nil
-	}
-	// check to see if the RollappId has been registered before with same key
+
+	// check to see if the RollappId has been registered before with same EIP155 ID
```
Comment (danwt) [inline x/rollapp/keeper/msg_server_create_rollapp.go:54]: useless comment

File: x/rollapp/keeper/msg_server_create_rollapp_test.go [hunk 3]
```diff
@@ -94,18 +91,58 @@ func (suite *RollappTestSuite) TestCreateRollappAlreadyExists() {
...
-		RollappId:             "rollapp1",
-		MaxSequencers:         1,
-		PermissionedAddresses: []string{},
+		Creator:   alice,
+		RollappId: "rollapp_1234-1",
 	}
+
 	_, err := suite.msgServer.CreateRollapp(goCtx, &rollapp)
-	suite.Require().Nil(err)
+	suite.Require().NoError(err)
```
Comment (danwt) [inline x/rollapp/keeper/msg_server_create_rollapp_test.go:100]: ❤️

File: x/rollapp/keeper/rollapp.go [hunk 4]
```diff
@@ -65,6 +62,24 @@ func (k Keeper) GetRollapp(
 	return val, true
 }
 
+func (k Keeper) FindRollappByName(
+	ctx sdk.Context,
```
Comment (danwt) [inline x/rollapp/keeper/rollapp.go:66]: GetRollappByName is more consistent func naming

File: app/apptesting/test_suite.go [hunk 5]
```diff
@@ -98,15 +100,23 @@ func (s *KeeperTestHelper) FundAcc(acc sdk.AccAddress, amounts sdk.Coins) {
...
+	s.Require().Equal(oldState, newState)
+}
+
+func GenerateRollappID() string {
+	name := make([]byte, 8)
+	for i := range name {
+		name[i] = byte(rand.Intn('z'-'a'+1) + 'a')
+	}
+	return fmt.Sprintf("%s_%d-1", string(name), rand.Int63())
 }
```
[the remark this answers is not in the document]
Comment (danwt) [reply app/apptesting/test_suite.go:122]: can you link the eip55 standadr somewhere too

File: x/rollapp/keeper/rollapp.go [hunk 6]
```diff
@@ -18,15 +19,12 @@ func (k Keeper) SetRollapp(ctx sdk.Context, rollapp types.Rollapp) {
 
 	// check if chain-id is EVM compatible. no err check as rollapp is already validated
 	rollappID, _ := types.NewChainID(rollapp.RollappId)
```
Comment (danwt) [inline x/rollapp/keeper/rollapp.go:21]: imo Must function is better here

File: x/rollapp/types/errors.go [hunk 7]
```diff
@@ -32,6 +32,7 @@ var (
 	ErrNotFound                       = errorsmod.Register(ModuleName, 1037, "not found")
 	ErrLogic                          = errorsmod.Register(ModuleName, 1038, "internal logic error")
 	ErrInvalidAddress                 = errorsmod.Register(ModuleName, 1040, "invalid address")
+	ErrRollappIDNotEIP155             = errorsmod.Register(ModuleName, 1041, "rollapp-id is not EIP155 compliant")
```
Comment (danwt) [inline x/rollapp/types/errors.go:35]: there is already ErrInvalidRollappID so this error should wrap it, not be a new one

**Review by danwt — CHANGES_REQUESTED** [review]
Nice cleanup in general
Left a few things

File: app/apptesting/test_suite.go [hunk 8]
```diff
@@ -42,7 +44,7 @@ func (s *KeeperTestHelper) CreateRollappWithName(name string) string {
 
 	msgServer := rollappkeeper.NewMsgServerImpl(*s.App.RollappKeeper)
 	_, err := msgServer.CreateRollapp(s.Ctx, &msgCreateRollapp)
-	s.Require().NoError(err)
+	s.Require().NoErrorf(err, "failed to create rollapp %s", name)
```
[the remark this answers is not in the document]
Comment (zale144) [reply app/apptesting/test_suite.go:47]: I guess I failed to remember that I should not have in errors 🙂

**Review by omritoptix — APPROVED** [review]
approved with small comments (about left-over godocs)

**Review by omritoptix — APPROVED** [review]

**Review by danwt — APPROVED** [review]
gj!

**Pull request merged**